src/server/managers/AppExternalComponentManager.ts

```diff
@@ -1,79 +1,129 @@
 import { IExternalComponent } from '../../definition/externalComponent';
+import { ExternalComponentAlreadyTouchedError, ExternalComponentNotMatchWithAppError } from '../errors';
 
 /**
- * The external component manager for the Apps.
+ * The external component manager for the apps.
  *
- * An App will register an external component during their `initialize` method.
- * Then once an App's `onEnable` is called and it returns true,
- * only then will that App's external component be enabled.
+ * An app will register external components during its `initialize` method.
+ * Then once an app's `onEnable` method is called and it returns true,
+ * only then will that app's external components be enabled.
  */
 export class AppExternalComponentManager {
     /**
-     * The map that maintain the list of all registered components.
+     * The map that maintains all registered components.
+     * The key of the top map is app id and the key of inner map is the
+     * external component name.
      */
-    private providedComponents: Map<string, IExternalComponent>;
+    private providedExternalComponents: Map<string, Map<string, IExternalComponent>>;
```

Comment:
Let's call this variable `registeredExternalComponents` and then the one below `registeredExternalComponents`. The reason the slash command uses the terminology `touched` is because we allow Apps to override system commands. But external components are new to the Apps and they can't overwrite other Apps ones or system ones.

Reply:
Resolved in https://github.com/RocketChat/Rocket.Chat.Apps-engine/pull/150/commits/bfc6fd3fedbcb85107a83e9b0e760727710a64fe